### ISSUE TITLE
Allow def symbols when using :allow whitelist with sci.core/init

### ISSUE THREAD

**Issue by teodorlu** [borkdude/sci#434]
**Is your feature request related to a problem? Please describe.**
I'd like to use an `:allow` whitelist for `sci.core/init`, but also allow to `def` arbitrary symbols.

**Describe the solution you'd like**
I can whitelist symbols to be used in evaluation:

```clojure
(let [ctx (sci/init {:allow ['* 'def '+]})]
  (sci/eval-form ctx '(+ 17 25)))
;; => 42
```

but I'd have manually allow each symbol bound with def:

```clojure
;; But I cannot know in advance which symbols a user will use in defs
(let [ctx (sci/init {:allow ['* 'def '+ 'do]})]
  (sci/eval-form ctx
                 '(do (def answer (+ 17 25))
                      (* 10 answer))))
;; 1. Unhandled clojure.lang.ExceptionInfo
;;    answer is not allowed!
;;    {:type :sci/error, :line nil, :column nil, :file nil, :phase "analysis"}
```

With `let`, arbitrary symbols are allowed:

```clojure
(let [ctx (sci/init {:allow ['* 'let '+ 'do]})]
  (sci/eval-form ctx
                 '(let [answer (+ 17 25)]
                    (* 10 answer))))
;; => 420
```

**Describe alternatives you've considered**
- Allowing anything and validating in advance. Requires to be able to control what code is generated.

**Possible solutions**
_Changing :allow semantics to allow the user to define own symbols_. Backwards-incompatible, possibly not desirable.

_Add new option to sci.core/init_. Example: `(sci.core/init {:allow '[def + -] :allow-def-symbols true)`.

- Question: should `defn`, `def` and `defmacro` be treated with _one_ option, or should they be possible to control individually?
- Question: is the value behind :allow-def-symbols a boolean or a keyword?

**My usecase specifically**
I'm building a language on top of Sci. I'd like to use the whitelist to be explicit about what subset of Sci I'm using. However, I need to support binding variables.

One option is to rewrite my code to a let. But then I would not have intermediate variables available in ctx after running.

Thanks for a great project!

**Comment by borkdude** [comment]
> Changing :allow semantics to allow the user to define own symbols. Backwards-incompatible, possibly not desirable.

I don't think anyone will have a problem with this and we can update accordingly. Sci vars have `:sci.impl/built-in` set to `true`, this allows checking if the var was introduced by a user.

Possibly related: https://github.com/borkdude/sci/issues/357

**Comment by borkdude** [comment]
@teodorlu Fixed with https://github.com/borkdude/sci/commit/60f02076bf700410f65ec2f817db407d8de2d172. Do you need a new Clojars version or are you using git deps?

**Comment by teodorlu** [comment]
Wow, thanks!

I'm using Clojars, so I'd love to see a new release.

> Allow user-defined vars when def is allowed

Would allowing `defn` or `defmacro` also cause user-defined vars to be allowed?

**Comment by borkdude** [comment]
Will be available as `0.1.1-alpha.9` in a few minutes. And yes.
